### PR TITLE
[rush] Move the rush-lib bundle to lib/index.js.

### PR DESCRIPTION
**Pull request by iclanton**
This fixes an issue that @chengcyber discovered during testing.

### CODE REVIEW

**Comment by octogonz** [comment]
> Please elaborate on what this is addressing. The naming convention gets a bit confusing if `lib/`, which normally contains unminified directly-transpiled modules instead contains the bundled code.

@dmichon-msft  The underlying problem is that `lib/start.js` is a [longstanding interface contract](https://github.com/microsoft/rushstack/blob/9360abffdf2066fa095ca056a319a0b10677fb1b/apps/rush/src/RushVersionSelector.ts#L86) for the `@microsoft/rush` version selector.  So if we start bundling the code, we need to preserve that entry point.  Also @chengcyber has pointed out that various scripts have relied on deep imports to access internal Rush APIs as a workaround for missing public APIs.  (Example: `Utilities.js`, `SelectionParameters.js`)

However, I agree that the approach in this PR is a bit counterintuitive. We either need to document it very well, or find a more straightforward approach (for example, copying everything into a [publishFolder](https://github.com/microsoft/rushstack/blob/9360abffdf2066fa095ca056a319a0b10677fb1b/rush.json#L384).

**Comment by iclanton** [comment]
See https://github.com/microsoft/rushstack/pull/3840

**Comment by octogonz** [comment]
@dmichon-msft and I chatted about this PR today, since Ian is away.  Our thinking:

Let's tackle this fix in two steps:  

**Step 1** will be a small PR that temporarily reverts the feature, by redirecting `"main"` in **package.json** to point to `lib/index.js` again

**Step 2** will reintroduce bundling with a modified approach:

- `rush-lib` will get published from a [publishFolder](https://github.com/microsoft/rushstack/blob/9360abffdf2066fa095ca056a319a0b10677fb1b/rush.json#L384), i.e. after compiling, the output files will get copied into a separate folder to postprocessing easier to understand

- The webpacked bundle will be `publish/dist/rush-lib.js` 
- The contract paths (`publish/lib/index.js`, `publish/lib/start.js`, `publish/lib/start-pnpm.js`, etc) will be stubs that load the appropriate export from `../dist/rush-lib.js`.
- David proposed that these stubs could be automatically generated by a custom Webpack plugin, which also generates stubs for every `src/*.ts` file.  This would be valuable for supporting (1) scripts that need to import Rush internals as a workaround to access missing public APIs and (2) Jest tests for Rush plugins that need to mock Rush internals.
- This might sacrifice the `async import` chunks ([example here](https://github.com/microsoft/rushstack/blob/3e0ec6eb9ef483d0fa3c6e4f28ed4229c90530ce/libraries/rush-lib/src/cli/scriptActions/PhasedScriptAction.ts#L409)), but maybe that optimization is overengineering given Rush's requirements.

**Comment by octogonz** [comment]
BTW there are other possibilities besides `async import`. Consider something like:

```ts
import { ProjectWatcher } from `rush-lib`;

async function doStuff(): Promise<void> {
  // Expensive loading of chunk/dependencies happens here
  await ProjectWatcher.ensureLoadedAsync();

  const projectWatcher = new ProjectWatcher();
  projectWatcher.doThing();
}
```

In this way the imports remain simple and synchronous, but the expensive code can still be loaded deferred/asynchronously.  Just an idea...

**Comment by iclanton** [comment]
@octogonz, @dmichon-msft see https://github.com/microsoft/rushstack/pull/3840 as a hacked-together example of what you're describing for Step 2

**Comment by iclanton** [comment]
@octogonz, updated this PR to cover Step 1.

File: libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export interface IPnpmfileShimOptions {
 export class PnpmfileConfiguration {
   private _context: IPnpmfileContext | undefined;
 
-  public constructor(context: IPnpmfileContext) {
+  private constructor(context: IPnpmfileContext) {
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/pnpm/PnpmfileConfiguration.ts:33]: This is included in this PR as an extra precaution to make sure the `__non_webpack_require__` function in `PnpmfileShim` never gets called.

**Review by octogonz — APPROVED** [review]
Thanks Ian!

**Pull request merged**